mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,171 @@
+        r"""
+        Returns a list of measurement samples on a specified state. If ``self.probabilities`` is
+        ``None`` then uses a uniform probability distribution.
+
+        Args:
+            state: The state to generate samples of.
+            n_samples: The number of samples to generate.
+        """
+        rng = settings.rng
+        return rng.choice(a=self._meas_outcomes, p=self.probabilities(state), size=n_samples)
+
+    def _validate_state(self, state: State | None = None):
+        r"""
+        Validates that the modes of ``state`` and ``self.meas_ops`` are compatible with one another.
+
+        Args:
+            state: The state to validate.
+        """
+        if self.meas_ops and state is not None:
+            meas_op_modes = (
+                self.meas_ops[0].modes
+                if isinstance(self.meas_ops, Iterable)
+                else self.meas_ops.modes
+            )
+            if not set(state.modes) >= set(meas_op_modes):
+                raise ValueError(
+                    f"State with modes {state.modes} is incompatible with the measurement operators with modes {meas_op_modes}."
+                )
+
+
+class PNRSampler(Sampler):
+    r"""
+    A sampler for photon-number resolving (PNR) detectors.
+
+    Args:
+        modes: The measured modes.
+        cutoff: The photon number cutoff.
+    """
+
+    def __init__(self, modes: Sequence[int], cutoff: int) -> None:
+        super().__init__(list(range(cutoff)), [Number(modes, n) for n in range(cutoff)])
+
+    def probabilities(self, state: State | None = None) -> list[float] | None:
+        fock_state = state.to_fock() if state else state
+        return super().probabilities(fock_state)
+
+
+class HomodyneSampler(Sampler):
+    r"""
+    A sampler for homodyne measurements.
+
+    Args:
+        modes: The measured modes.
+        phi: The quadrature angle where ``0`` corresponds to ``x`` and ``\pi/2`` to ``p``.
+        bounds: The range of values to discretize over.
+        num: The number of points to discretize over.
+    """
+
+    def __init__(
+        self,
+        modes: Sequence[int],
+        phi: float = 0,
+        bounds: tuple[float, float] = (-5, 5),
+        num: int = 100,
+    ) -> None:
+        super().__init__(list(np.linspace(*bounds, num)), BtoQ(modes, phi=phi))
+
+    def probabilities(self, state: State | None = None):
+        self._validate_state(state)
+        if state is not None:
+            disjoint_modes = [mode for mode in state.modes if mode not in self.meas_ops.modes]
+            dm_state = state.dm() >> TraceOut(disjoint_modes) if disjoint_modes else state.dm()
+            q_state = dm_state >> self.meas_ops
+            z = [[x] * q_state.representation.ansatz.num_vars for x in self.meas_outcomes]
+            probs = math.real(q_state.representation(z)) * math.sqrt(settings.HBAR)
+            return probs / sum(probs)
```

Comment:
Like above it could be good to validate a valid probability distribution.  This would require here multiplying the probability density by the differential (dx), but that should be possible as you have the whole list of points and the differential could be estimated by the difference between points.

Also, is `math.real` required here?  Shouldn't that be part of the `representation` code?

Reply:
Yeah I agree that there should be something for validation. I had some form of it before but was getting annoying rounding errors so I ended up just doing the lazy `/ sum(probs)`. I'll play around with it some more to see if I can get something working .

Unfortunately, the representation gives complex values so I think the `math.real` is required (or at least nice to have).

Reply:
Perhaps this is a case of having to specify that you want this re-normalisation though an argument.  Lots of weird behaviour can be expected from probability distributions that are conditional (i.e. here conditional on the events that you keep).  Looking at the numpy source for `choice`, the total probability requirements are very strict.  The probabilities must sum to 1.0 within the sqrt of machine epsilon for the float type being used.  If you do adopt an argument, it might be good to accept a custom absolute tolerance as we generally accept distributions that are within 1e-4 of 1.0.

Reply:
I've managed to get a probability validation method on the Sampler. Turns out some of the probability computations I was doing was neglecting terms in the multimode case causing it to fail. I've updated PNRSampler working on the Homodyne now